File: kadaneAlgo.py
'''
Maximum_subarray_problem

Summary:
Given an array, it will computer the value
of the largest subarray, it doesn't keep track
of the indices though
'''
def max_sub(Array):
    
    #set default values of accumulators to first element
    max_ending_here = max_so_far = Array[0]
    
    #iterate through array
    for val in Array:
        
        #get the ending_here accumulator and add val
        #if the new value is negative, set it to 0
        max_ending_here = max(0, max_ending_here + val)
        
        #get the bigger value out of the accumulator so far
        #and the ending_here value
        max_so_far = max(max_so_far, max_ending_here)
        
    #the value returned will be the maximum sum
    #it found that's bigger than 0
    return max_so_far

ranArray = [-1,2,3,-4,5,6,7,-3,-1]
max_subarray(ranArray)

File: kadaneAlgo.java

public class test {

	/**
	 * Main method
	 * @param args
	 */
	public static void main(String[] args) {
		int[] array = new int[]{-1,2,3,-4,5,6,7,-3,-1};
		System.out.println(maxSub(array));
	}
	
	/**
	 * Kadane's algorithm for solving
	 * the maximum subarray problem
	 * @param array of integers
	 * @return value of maximum subarray
	 */
	public static int maxSub(int[] array) {
		int maxSoFar = 0;
		int maxEndingHere = 0;
		int temp;
		
		for(int i: array){
			temp = i + maxEndingHere;
			maxEndingHere = (temp > 0)? temp : 0;
			maxSoFar = (maxEndingHere > maxSoFar)? maxEndingHere : maxSoFar;
		}
		
		return maxSoFar;
	}

}
